Adding commands to jump to first/last row/column in tables

mltony opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft. See https://github.com/nvaccess/nvda/wiki/Contributing
-->

### Link to issue number:
fixes #957

### Summary of the issue:
Introduce commands to jump to first/last row/column in current table.
### Description of how this pull request fixes the issue:
Implements new commands in class DocumentWithTableNavigation.
### Testing strategy:
* Tested in Chrome, Firefox, Edge.
* Tested in HTML tables with merged cells provided in https://github.com/nvaccess/nvda/pull/13345#issuecomment-1046376381
* Also tested on MSWord table from previous comment in both browse and focus mode.
* Tested in HTML table with missing cells.
### Known issues with pull request:
Will cause merge conflicts with #13345 - will be happy to resolve them once it lands.
### Change log entries:
New features
* New table navigation commands: Control+Alt+Home/End to jump to first/last column; Control+Alt+PageUp/PageDown to jump to first/last row.
Changes
Bug fixes
For Developers

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

mltony committed:
Adding commands to jump to first/last row/column in tables

mltony commented:
cc: @feerrenrut, @seanbudd

lukaszgo1 commented on source/documentBase.py:
```diff
@@ -86,6 +86,15 @@ def _getTableCellAt(self,tableID,startPos,row,column):
+		Returns height and width of the table.
+		@param tableID: the ID of the table.
+		@returns: Height and width of the table.
+		@rtype: L{Tuple[int, int]}
```
For new code types for parameters and return values should be specified using Python type annotations if possible. The same goes for all docstrings below - I'm not going to comment on all of them.

mltony replied:
Fixed that. The only thing I couldn't annotate is table_id as it is a union. This union is properly defined in #13345 - will annotate once it lands.

lukaszgo1 commented on source/documentBase.py:
```diff
@@ -138,6 +147,94 @@ def _getNearestTableCell(self, tableID, startPos, origRow, origCol, origRowSpan,
+		@param origCol: the column number  of the starting cell
+		@type origCol: int
+		@param axis: the axis of movement ("row" or "column")
+		@type axis: string
```
While this should not block this PR ideally both axes and directions  of movement should be specified as an enum.

mltony replied:
Already implemented in #13345.

lukaszgo1 commented on source/documentBase.py:
```diff
@@ -138,6 +147,94 @@ def _getNearestTableCell(self, tableID, startPos, origRow, origCol, origRowSpan,
+		"""
+		result = origRow if axis == "row" else origCol
+		row, col = origRow, origCol
+		maxTableSize = 1000
```
A comment here explaining how you come with the 1000 as a right value would be nice.

mltony replied:
This is pretty much an arbitrary large number. I added a pretty obvious comment. There is no science behind it.

lukaszgo1 commented on source/documentBase.py:
```diff
@@ -189,7 +289,22 @@ def script_previousColumn(self, gesture):
-
+	def script_firstRow(self, gesture):
+		self._tableMovementScriptHelper(axis="row", movement="first")
+	# Translators: the description for the first table row script on browseMode documents.
```
For new scripts script decorator should be used to set script's metadata. Obviously for `documentBase` this cannot be done at present as it would reintroduce #12399, but perhaps an issue should be created to at least avoid this problem being forgotten.

lukaszgo1 commented on source/virtualBuffers/gecko_ia2.py:
```diff
@@ -509,6 +509,18 @@ def _getTableCellAt(self,tableID,startPos,destRow,destCol):
+				table.IAccessibleTable2Object.nRows,
+				table.IAccessibleTable2Object.nColumns,
+			)
+		except OSError:#(COMError, RuntimeError, AttributeError):
```
I don't see how `OSError` may be raised by this code. Could you elaborate? At a guess this should be a `COMError`.

mltony replied:
Good catch! Deleted OSError.

lukaszgo1 reviewed (changes requested):
Please add  the newly introduced gestures to the user guide.

mltony commented on source/documentBase.py:
```diff
@@ -189,7 +289,22 @@ def script_previousColumn(self, gesture):
+	def script_firstRow(self, gesture):
+		self._tableMovementScriptHelper(axis="row", movement="first")
+	# Translators: the description for the first table row script on browseMode documents.
+	script_firstRow.__doc__ = _("moves to the first table row")	
```
I don't have the background on why this would be impossible, so I don't feel I can create a meaningful issue.

mltony commented:
@lukaszgo1,  Addressed your comments.

seanbudd commented:
The build is failing as linting needs to be fixed.

Adriani90 commented:
@mltony can the commands be reassigned by the user? Me personally for example I would prefer to use ctrl+alt+nvda+arrow keys to jump to last column or row. The thing is that my laptop keyboard does not have home, end, page up and page down keys, so I had to replace some f keys with sharp keys to acomplish this function. But is quite unconfortable to use them very often.

CyrilleB79 commented on user_docs/en/changes.t2t:
```diff
@@ -23,6 +23,7 @@ What's New in NVDA
 - Added commands for toggling multiple modifiers simultaneously with a Braille display (#13152)
 - The Speech Dictionary dialog now features a "Remove all" button to help clear a whole dictionary. (#11802)
 - Added support for Windows 11 Calculator. (#13212)
+- Added new commands to jump to first/last row/column in tables.
```
Please remove the change on this file. To avoid dealing with many merge conflicts on the change log, NVAccess prefers to update this file themselves. Once your PR is accepted, they will look at the change log section in the initial description of the PR and they will copy/paste its content to this file.

CyrilleB79 reviewed (commented):
Hi Tony

Really happy to see this PR arriving in core!

Some comments here:
* You should test (and implement if needed) the same type of commands in list views where table navigation commands are also available.
  If this is not possible for any reason, you should document it and explain why in this PR's initial description.
  Note: IMO, if this is not possible for list views, it would already be a great improvement to merge this PR only for real tables.. But it must be clearly documented.
* Testing: For Word, did you test with and without UIA enabled? Please update the test section in the initial description accordingly.

Thanks!

Qchristensen reviewed (approved):
User Guide change is good, thanks - good work!

michaelDCurran reviewed (changes requested):
It is possible to write code to get the dimensions of the table without having to resort to hit testing to find the limits.
From any given textInfo position, you can expand to character, getTextWithFields,  walk up the ancestor controlFields until reaching the inner most table controlField, and then fetch table-rowcount and table-columncount attributes.
Refer to `_getTableCellCoords` for a similar example of where attributes are fetched off a table controlField.

mltony commented:
@seanbudd, fixed lint and build.
@Adriani90, no problem to reassign keystrokes.
@CyrilleB79, codepath is entirely different for list views (it is implemented in RowWithFakeNavigation class IIRC), so I'd rather separate that into another PR. Also, for word, I tried turning UIA on and off for word in NVDA settings, and it seems to be still  using UIA even when turned off. But now with Michael's suggestion I feel this is irrelevant since we don't need to interact with low-level IA2/UIA API.
@michaelDCurran, thanks, changed code according to your suggestion. Also refactored a bit to avoid code reduplication.
The bot says system tests failed, but from what I can see, it appears that only 1 test is broken and it's probably transient and does not appear to be related:
```
[00:24:52] Starts from desktop shortcut :: Ensure that NVDA can start from de... | FAIL |
[00:24:52] Connection to remote server broken: [WinError 10061] No connection could be made because the target machine actively refused it
```

mltony commented:
Also PR #13470 provides same keystrokes in list views and tree views.

seanbudd commented:
@mltony - can you resolve the merge conflicts as you mentioned in the PR description. I've just merged #13345

mltony commented:
@seanbudd, resolved conflicts!

seanbudd commented on source/documentBase.py:
```diff
@@ -169,12 +221,68 @@ def _getNearestTableCell(
+			origCol: int,
+			origRowSpan: int,
+			origColSpan: int,
+			movement: str,
+			axis: str
```
```suggestion
			movement: _Movement,
			axis: _Axis
```

seanbudd commented on source/documentBase.py:
```diff
@@ -169,12 +221,68 @@ def _getNearestTableCell(
+		@param origCol: the column number  of the starting cell
+		@param origRowSpan: the row span of the row of the starting cell
+		@param origColSpan: the column span of the column of the starting cell
+		@param movement: the direction ("next" or "previous")
```
```suggestion
		@param movement: the direction ("first"/"next" or "last"/"previous")
```

seanbudd reviewed (commented):
LGTM, just some changes on documentation / types

mltony commented:
@seanbudd, fixed and rebased.

seanbudd reviewed (approved):
thanks @mltony.

please avoid rebases in future for minor changes if possible. After a rebase, GitHub no longer allows you to review pushed changes since your last review. This means the entire PR must be re-reviewed, instead of the changes.

seanbudd committed:
Merge branch 'master' into table_last

seanbudd committed:
Update changes
